File: src/main/java/indi/fcwyzzr/minecraft/food_talks/common/data_component/SimpleDataComponents.java
package indi.fcwyzzr.minecraft.food_talks.common.data_component;

import com.mojang.serialization.Codec;
import indi.fcwyzzr.minecraft.food_talks.common.data_component.compound_food.FoodItemProperties;
import net.minecraft.core.Holder;
import net.minecraft.core.NonNullList;
import net.minecraft.core.component.DataComponentType;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.food.FoodProperties;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;

import java.util.List;

public interface SimpleDataComponents {
    interface SimpleCodecs {
        Codec<List<Holder<Item>>> itemList = ItemStack.ITEM_NON_AIR_CODEC.listOf();
        Codec<List<ItemStack>> itemStackList = ItemStack.SINGLE_ITEM_CODEC.listOf();
        Codec<List<FoodProperties.PossibleEffect>> possibleEffectList =
            FoodProperties.PossibleEffect.CODEC.listOf();
    }
    interface SimpleStreamCodecs{
        StreamCodec<RegistryFriendlyByteBuf, List<Holder<Item>>> itemList = ByteBufCodecs
            .holderRegistry(Registries.ITEM).apply(
                ByteBufCodecs.collection(NonNullList::createWithCapacity)
            );
        StreamCodec<RegistryFriendlyByteBuf, List<ItemStack>> itemStackList = ItemStack.LIST_STREAM_CODEC;
        StreamCodec<RegistryFriendlyByteBuf, List<FoodProperties.PossibleEffect>> possibleEffectList =
            FoodProperties.PossibleEffect.STREAM_CODEC.apply(
                ByteBufCodecs.collection(NonNullList::createWithCapacity)
            );
        
    }
    
    DataComponentType<List<Holder<Item>>> SandwichLayer = DataComponentType
        .<List<Holder<Item>>>builder()
        .persistent(SimpleCodecs.itemList)
        .networkSynchronized(SimpleStreamCodecs.itemList)
        .build();
    
    DataComponentType<List<FoodProperties.PossibleEffect>> PossibleEffectList =
        DataComponentType
            .<List<FoodProperties.PossibleEffect>>builder()
            .persistent(SimpleCodecs.possibleEffectList)
            .networkSynchronized(SimpleStreamCodecs.possibleEffectList)
            .build();
}

File: src/main/kotlin/indi/muxin/food_talks/common/data_components/SimpleDataComponents.kt
package indi.muxin.food_talks.common.data_components

import indi.muxin.food_talks.toRegistryName
import indi.muxin.food_talks.toResourceLocation
import net.minecraft.core.Holder
import net.minecraft.core.NonNullList
import net.minecraft.core.component.DataComponentType
import net.minecraft.core.registries.Registries
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.world.food.FoodProperties.PossibleEffect
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.neoforged.neoforge.registries.RegisterEvent
import java.util.function.IntFunction

val SandwichLayerDCType: DataComponentType<List<Holder<Item>>> = DataComponentType
    .builder<List<Holder<Item>>>()
    .persistent(ItemStack.ITEM_NON_AIR_CODEC.listOf())
    .networkSynchronized(ByteBufCodecs
        .holderRegistry(Registries.ITEM).apply(
            ByteBufCodecs.collection { NonNullList.createWithCapacity(it) }))
    .build()

val PossibleEffectListDCType: DataComponentType<List<PossibleEffect>> = DataComponentType
    .builder<List<PossibleEffect>>()
    .persistent(PossibleEffect.CODEC.listOf())
    .networkSynchronized(PossibleEffect.STREAM_CODEC.apply(
        ByteBufCodecs.collection(IntFunction { NonNullList.createWithCapacity(it) })))
    .build()

fun RegisterEvent.RegisterHelper<DataComponentType<*>>.registerSimpleDataComponents() {
    register("SandwichLayer".toRegistryName().toResourceLocation(), SandwichLayerDCType)
    register("PossibleEffectList".toRegistryName().toResourceLocation(), PossibleEffectListDCType)
}